### DISCUSSION TITLE
How to prevent vnode properties prefixed with underline from being renamed ?

### DISCUSSION THREAD

**Discussion by RRDAWLX** [preactjs/preact#3266] [Q&A] [answered]
The source codes of vnode are like these:
```
const vnode = {
    type,
    props,
    key,
    ref,
    _children: null,
    _parent: null,
    _depth: 0,
    _dom: null,
    // _nextDom must be initialized to undefined b/c it will eventually
    // be set to dom.nextSibling which can return `null` and it is important
    // to be able to distinguish between an uninitialized _nextDom and
    // a _nextDom that has been set to `null`
    _nextDom: undefined,
    _component: null,
    _hydrating: null,
    constructor: undefined,
    _original: original == null ? ++vnodeId : original
};
```
But I find all the properties prefixed with underline are renamed to unreadable ones. Like these:
```
var vnode = {
    type: type,
    props: props,
    key: key,
    ref: ref,
    __k: null,
    __: null,
    __b: 0,
    __e: null,
    // _nextDom must be initialized to undefined b/c it will eventually
    // be set to dom.nextSibling which can return `null` and it is important
    // to be able to distinguish between an uninitialized _nextDom and
    // a _nextDom that has been set to `null`
    __d: undefined,
    __c: null,
    __h: null,
    constructor: undefined,
    __v: original == null ? ++vnodeId : original
};
```
It is very inconvenient for me to learn the implementation of preact. How can I turn off the renaming feature?

**Comment by JoviDeCroock** [comment, marked as answer]
This file [here](https://github.com/preactjs/preact/blob/master/mangle.json) contains the mapping of said properties, you can build preact with `--no-compress` if you want to see it without but essentially these are internals and we choose (for library size) not to expose these to the outside world. If we'd keep the original names in there Preact would be a few (50-100'ish) bytes bigger.

I guess learning the implementation is better done through the [source code](https://github.com/preactjs/preact/tree/master/src) as opposed to actual built code, as built code is tailored towards being optimized which can introduce unreadable parts.

**Reply by RRDAWLX** [reply]
Thanks. I can't understand all the implementations and mechanisms by reading source code. So I want to learn by runtime.
